execution_engine/src/storage/trie_store/operations/mod.rs

```diff
@@ -254,31 +254,25 @@ where
         if !visited.insert(trie_key) {
             continue;
         }
-        let maybe_retrieved_trie_bytes = store.get_raw(txn, &trie_key)?;
-        // Optimization: Don't look for descendants of leaves.
-        match maybe_retrieved_trie_bytes.and_then(|bytes| bytes.get(0)) {
-            // if this is a leaf don't parse (ie, the first byte is 0), just continue
-            Some(&Trie::<K, V>::LEAF_TAG) => continue,
-            // Either no entry under trie key or data is empty.  If data is empty then the trie is
-            // corrupted. Either way treat as missing.
+
+        let retrieved_trie_bytes = match store.get_raw(txn, &trie_key)? {
+            Some(bytes) => bytes,
             None => {
+                // No entry under this trie key.
                 missing_descendants.push(trie_key);
                 continue;
             }
-            _ => (),
+        };
+
+        // Optimization: Don't deserialize leaves as they have no descendants.
+        if let Some(&Trie::<K, V>::LEAF_TAG) = retrieved_trie_bytes.first() {
```

Comment:
(Now there's no need to match anymore.)

```suggestion
        if Some(&Trie::<K, V>::LEAF_TAG) == retrieved_trie_bytes.first() {
```

Reply:
Or there the case when `Bytes` are empty could be handled together with non-existence of `trie_key` in the trie. Does it make a difference whether it's not in the trie or the bytes are empty? i.e.
```rust
None | Some(bytes) if bytes.is_empty() => {
               // No entry under this trie key.
                missing_descendants.push(trie_key);
                continue;
}
```

Reply:
Not sure. I think I'd rather consider an existing but empty entry in the database to be the same as "can't deserialize", not "doesn't exist".

Reply:
But "can't deserialize" and "doesn't exist" are treated the same way - with the exeception of logging an error. Just compare the `None` branch above and the `Err` case below.

Reply:
Ah, right, then your suggestion is fine.
(Not sure if you can have an `if` in a `match` arm that applies only to one of the options separated by `|`?)

Reply:
Since Fraser is on holidays this week, I will merge this PR now and put up a following one with the suggested changes.